libsel4/tools/generator_rust.py

```diff
@@ -0,0 +1,604 @@
+            ],
+
+            "x86_64": [
+                Type("seL4_X86_VMAttributes", wordsize, wordsize),
+                CapType("seL4_X86_IOPort", wordsize),
+                CapType("seL4_X86_IOPortControl", wordsize),
+                CapType("seL4_X86_ASIDControl", wordsize),
+                CapType("seL4_X86_ASIDPool", wordsize),
+                CapType("seL4_X86_IOSpace", wordsize),
+                CapType("seL4_X86_Page", wordsize),
+                CapType("seL4_X64_PML4", wordsize),
+                CapType("seL4_X86_PDPT", wordsize),
+                CapType("seL4_X86_PageDirectory", wordsize),
+                CapType("seL4_X86_PageTable", wordsize),
+                CapType("seL4_X86_IOPageTable", wordsize),
+                CapType("seL4_X86_VCPU", wordsize),
+                CapType("seL4_X86_EPTPML4", wordsize),
+                CapType("seL4_X86_EPTPDPT", wordsize),
+                CapType("seL4_X86_EPTPD", wordsize),
+                CapType("seL4_X86_EPTPT", wordsize),
+                StructType("seL4_VCPUContext", wordsize * 7, wordsize),
+                StructType("seL4_UserContext", wordsize * 20, wordsize),
+            ],
+            "riscv32": [
+                Type("seL4_RISCV_VMAttributes", wordsize, wordsize),
+                CapType("seL4_RISCV_Page", wordsize),
+                CapType("seL4_RISCV_PageTable", wordsize),
+                CapType("seL4_RISCV_ASIDControl", wordsize),
+                CapType("seL4_RISCV_ASIDPool", wordsize),
+                StructType("seL4_UserContext", wordsize * 32, wordsize),
+            ],
+            "riscv64": [
+                Type("seL4_RISCV_VMAttributes", wordsize, wordsize),
+                CapType("seL4_RISCV_Page", wordsize),
+                CapType("seL4_RISCV_PageTable", wordsize),
+                CapType("seL4_RISCV_ASIDControl", wordsize),
+                CapType("seL4_RISCV_ASIDPool", wordsize),
+                StructType("seL4_UserContext", wordsize * 32, wordsize),
+            ]
+        }
+        # legacy alias
+        arch_types["arm_hyp"] = arch_types["aarch32"]
+        return arch_types
+
+    @staticmethod
+    def generate_param_list(input_params, output_params):
+        # Generate parameters
+        params = []
+        for param in input_params:
+            if not param.type.pass_by_reference():
+                params.append(param.type.render_parameter_name(param.name))
+            else:
+                params.append(param.type.pointer(False).render_parameter_name(param.name))
+        for param in output_params:
+            if param.type.pass_by_reference():
+                params.append(param.type.pointer(True).render_parameter_name(param.name))
+
+        return ", ".join(params)
+
+    def _gen_file_header(self):
+        # nothing needed for Rust
+        return
+
+    def _gen_file_footer(self):
+        # nothing needed for Rust
+        return
+
+    def _gen_conditional_compilation(self, condition) -> str:
```

Comment:
Seems the follow up taks here could be making the xml a bit more C preprocessor agostic. It would just give a commpa-separated list of `[!]CONFIG_xxx` tokens. We only have "A and B" conditions. If `CONFIG_MAX_NUM_NODES > 1` is replaced by `CONFIG_SMP_SUPPORT` then also, parsing should be simple in each generator to create the Rust/C specific line.

Reply:
OK, seems this conditinal stuff is a rabbit hole. In https://github.com/axel-h/seL4/tree/patch-axel-74 I see that changing to `CONFIG_SMP_SUPPORT` causes some failures. And there are more script that use the xml's `conditional` statement, so using a C-preprocessors agnostic languate requires changing them also. Seems simpler to ensure this uses a common syntax everywhere and then build a small parser for the Rust generator that rejects everything it does not understand.